### PR TITLE
Refactor grain state serialization

### PR DESCRIPTION
**Pull request by wassim-k**
I've refactored grain state serialization to follow the same pattern used by built-in storage providers in order to make the experience uniform for users of the library.
This refactor allows developers to customize the serializer used per storage provider while still providing easy to use defaults.

### CODE REVIEW

File: Orleans.Providers.MongoDB/StorageProviders/Serializers/BsonGrainStateSerializer.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        }
+
+        public BsonValue Serialize<T>(T state)
+        {
+            return BsonDocumentWrapper.Create(state);
+        }
+
+        public T Deserialize<T>(BsonValue value)
+        {
+            if (value.IsBsonDocument)
```
[the remark this answers is not in the document]
Comment (SebastianStehle) [reply Orleans.Providers.MongoDB/StorageProviders/Serializers/BsonGrainStateSerializer.cs:22]: I guess the serializer does only supports objects at the root level.

File: Orleans.Providers.MongoDB/StorageProviders/Serializers/Configuration/IStateProviderSerializerOptions.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Orleans.Providers.MongoDB.StorageProviders.Serializers.Configuration
+{
+    public interface IStateProviderSerializerOptions
```
Comment (SebastianStehle) [inline Orleans.Providers.MongoDB/StorageProviders/Serializers/Configuration/IStateProviderSerializerOptions.cs:3]: What is this interface needed for?

File: Orleans.Providers.MongoDB/StorageProviders/Serializers/IGrainStateSerializer.cs [hunk 3]
```diff
@@ -1,11 +1,11 @@
-﻿using Newtonsoft.Json.Linq;
+﻿using MongoDB.Bson;
 
-namespace Orleans.Providers.MongoDB.StorageProviders
+namespace Orleans.Providers.MongoDB.StorageProviders.Serializers
 {
     public interface IGrainStateSerializer
```
Comment (SebastianStehle) [inline Orleans.Providers.MongoDB/StorageProviders/Serializers/IGrainStateSerializer.cs:5]: Shall we not document every public type and method?

File: Orleans.Providers.MongoDB/StorageProviders/Serializers/JsonGrainStateSerializer.cs [hunk 4]
```diff
@@ -10,34 +12,22 @@ public class JsonGrainStateSerializer : IGrainStateSerializer
     {
         private readonly JsonSerializer serializer;
 
-        public JsonGrainStateSerializer(IServiceProvider serviceProvider, MongoDBGrainStorageOptions options)
+        public JsonGrainStateSerializer(IOptions<JsonGrainStateSerializerOptions> options, IServiceProvider serviceProvider)
         {
             var jsonSettings = OrleansJsonSerializerSettings.GetDefaultSerializerSettings(serviceProvider);
```
Comment (SebastianStehle) [inline Orleans.Providers.MongoDB/StorageProviders/Serializers/JsonGrainStateSerializer.cs:17]: Can we not inject the settings?

File: Test/Grains/BsonGrain.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Threading.Tasks;
+using Orleans.Providers.MongoDB.Test.GrainInterfaces;
+using Orleans.Runtime;
+
+namespace Orleans.Providers.MongoDB.Test.Grains
+{
+    public sealed class BsonGrainGrain : IBSonGrain
```
Comment (SebastianStehle) [inline Test/Grains/BsonGrain.cs:8]: Casing error? BSon instead of Bson

**Review by SebastianStehle — CHANGES_REQUESTED** [review]

File: Orleans.Providers.MongoDB/StorageProviders/Serializers/BinaryGrainStateSerializer.cs [hunk 6]
```diff
@@ -1,26 +1,29 @@
-﻿using Newtonsoft.Json.Linq;
+﻿using MongoDB.Bson;
+using Orleans.Serialization;
 using Orleans.Storage;
 
 namespace Orleans.Providers.MongoDB.StorageProviders.Serializers
 {
     public sealed class BinaryGrainStateSerializer : IGrainStateSerializer
     {
+        private const string BinaryElementName = "statedata";
```
[the remark this answers is not in the document]
Comment (wassim-k) [reply Orleans.Providers.MongoDB/StorageProviders/Serializers/BinaryGrainStateSerializer.cs]: Updated to `data`, to match azure table storage implementation.

File: Orleans.Providers.MongoDB/StorageProviders/Serializers/BsonGrainStateSerializer.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (wassim-k) [reply Orleans.Providers.MongoDB/StorageProviders/Serializers/BsonGrainStateSerializer.cs:22]: That's right, comment added.

File: Orleans.Providers.MongoDB/StorageProviders/Serializers/Configuration/IStateProviderSerializerOptions.cs (hunk 2 above)
Comment (wassim-k) [reply Orleans.Providers.MongoDB/StorageProviders/Serializers/Configuration/IStateProviderSerializerOptions.cs]: It's supposed to serve as a common interface for all grain storage providers that need GrainStateSerializer, but at a second look, we only have one storage provider, so I was able to remove it.
I've replaced it with `MongoDBGrainStorageConfigurator` which sets `GrainStateSerializer` value to the one registered in the DI under the same name as the provider.

File: Orleans.Providers.MongoDB/StorageProviders/Serializers/IGrainStateSerializer.cs (hunk 3 above)
Comment (wassim-k) [reply Orleans.Providers.MongoDB/StorageProviders/Serializers/IGrainStateSerializer.cs:8]: Updated.

File: Orleans.Providers.MongoDB/StorageProviders/Serializers/JsonGrainStateSerializer.cs (hunk 4 above)
Comment (wassim-k) [reply Orleans.Providers.MongoDB/StorageProviders/Serializers/JsonGrainStateSerializer.cs:17]: Can you elaborate on that please?

File: Test/Grains/BsonGrain.cs (hunk 5 above)
Comment (wassim-k) [reply Test/Grains/BsonGrain.cs]: Good catch, fixed.

File: Orleans.Providers.MongoDB/StorageProviders/Serializers/JsonGrainStateSerializer.cs (hunk 4 above)
Comment (wassim-k) [reply Orleans.Providers.MongoDB/StorageProviders/Serializers/JsonGrainStateSerializer.cs:17]: If you're referring to `MongoDBGrainStorageOptions`, it is now agnostic to the kind of serializer used, which is why `ConfigureJsonSerializerSettings` property has been replaced by `GrainStateSerializer` and `JsonGrainStateSerializerOptions` is used as for configuring `JsonGrainStateSerializer`.

Comment (SebastianStehle) [reply Orleans.Providers.MongoDB/StorageProviders/Serializers/JsonGrainStateSerializer.cs:17]: I was referring to ServiceProvider. But it is fine for me

**Comment by wassim-k** [comment]
We missed the version bump 🤦

**Pull request merged**